cmd/controller/main.go

```diff
@@ -34,24 +38,49 @@ import (
 	"github.com/google/knative-gcp/pkg/reconciler/messaging/channel"
 	"github.com/google/knative-gcp/pkg/reconciler/trigger"
 	"github.com/google/knative-gcp/pkg/utils/appcredentials"
+	"knative.dev/pkg/injection"
 	"knative.dev/pkg/injection/sharedmain"
+	"knative.dev/pkg/signals"
 )
 
 func main() {
 	appcredentials.MustExistOrUnsetEnv()
-	sharedmain.Main("controller",
-		auditlogs.NewController,
-		storage.NewController,
-		scheduler.NewController,
-		pubsub.NewController,
-		build.NewController,
-		staticpullsubscription.NewController,
-		kedapullsubscription.NewController,
-		topic.NewController,
-		channel.NewController,
+	ctx := signals.NewContext()
```

Comment:
I can't call this. It can only be called once per process, the second call immediately panics. `sharedmain.Main` calls it internally.

So, i think our choices are to either create singletons for the IAM client and IAM policy manager. And have them lazily initialize, similar to how gcpauth.StoreSingleton is being initialized. But at that point, is it better to use wire than just using a static singleton variable, like `iam.PolicyManagerSingleton`?

The other thing we could do is start altering the pkg code to either allow passing in constructor constructors (`func(context.Context, configmap.Watcher) injection.ControllerConstructor`) or a way to create the context (and configmap.Watcher) before creating the constuctor list for sharedmain.Main. My preference is the latter, but I don't know how knative will feel about that change. Thoughts?

Reply:
Can't this be fixed by called sharedmain.MainWithContext instead?

I had been working on a branch to generate wire provider sets to perform informer injection along with a refactored controller struct which would allow the entire controller be injected by wire. I haven't spent the time to finish this work up however, as it is a fairly significant divergence from the ctx based injection that is baked into much of pkg. If you think this sounds worthwhile I can dedicate some time to getting that PR out.

Reply:
Yes it can 😄